File: matrix_class/src/Doolittle.cpp
#include "../include/LUDecomposition.hpp"
#include <cmath>

#ifdef USE_CUDA
#include "../cuda/include/gpu_backend.cuh"
#include "../cuda/include/gpu_dispatch.cuh"
#endif

using namespace std;

// Doolittle constructors
Doolittle::Doolittle() : LUDecomposition() {}
Doolittle::Doolittle(int r, int c) : LUDecomposition(r, c) {}

// Doolittle LU decomposition
// L has 1s on diagonal, U is upper triangular
SolverResult Doolittle::solve(double *b, int n, int maxIter, double tol) {
  if (rows != cols || rows != n)
    throw MatrixException("matrix dimensions don't match for Doolittle LU");

  double **L = new double *[n];
  double **U = new double *[n];
  for (int i = 0; i < n; i++) {
    L[i] = new double[n];
    U[i] = new double[n];
    for (int j = 0; j < n; j++) {
      L[i][j] = 0.0;
      U[i][j] = 0.0;
    }
    L[i][i] = 1.0;
  }

#ifdef USE_CUDA
  if (BackendDispatcher::shouldUseGPU(n, "lu_factorize")) {
    double *flatA = new double[n * n];
    double *flatL = new double[n * n];
    double *flatU = new double[n * n];
    for (int i = 0; i < n; i++)
      for (int j = 0; j < n; j++) {
        flatA[i * n + j] = data[i][j];
        flatL[i * n + j] = L[i][j];
        flatU[i * n + j] = U[i][j];
      }
    gpuLU_Doolittle(flatA, flatL, flatU, n);
    for (int i = 0; i < n; i++)
      for (int j = 0; j < n; j++) {
        L[i][j] = flatL[i * n + j];
        U[i][j] = flatU[i * n + j];
      }
    delete[] flatA;
    delete[] flatL;
    delete[] flatU;
  } else {
#endif
    for (int k = 0; k < n; k++) {
      for (int j = k; j < n; j++) {
        double sum = 0.0;
        for (int s = 0; s < k; s++)
          sum += L[k][s] * U[s][j];
        U[k][j] = data[k][j] - sum;
      }
      for (int i = k + 1; i < n; i++) {
        double sum = 0.0;
        for (int s = 0; s < k; s++)
          sum += L[i][s] * U[s][k];
        if (fabs(U[k][k]) < 1e-12)
          throw MatrixException(
              "zero pivot in Doolittle... matrix might be singular bro");
        L[i][k] = (data[i][k] - sum) / U[k][k];
      }
    }
#ifdef USE_CUDA
  }
#endif

  // LU verification: compute max |L*U - A|
  double maxError = 0.0;
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < n; j++) {
      double sum = 0.0;
      for (int k = 0; k < n; k++)
        sum += L[i][k] * U[k][j];
      double err = fabs(sum - data[i][j]);
      if (err > maxError)
        maxError = err;
    }
  }

  double *y = new double[n];
  for (int i = 0; i < n; i++) {
    double sum = 0.0;
    for (int j = 0; j < i; j++)
      sum += L[i][j] * y[j];
    y[i] = b[i] - sum;
  }

  double *x = new double[n];
  for (int i = n - 1; i >= 0; i--) {
    double sum = 0.0;
    for (int j = i + 1; j < n; j++)
      sum += U[i][j] * x[j];
    if (fabs(U[i][i]) < 1e-12)
      throw MatrixException("zero on U diagonal in back sub... rip");
    x[i] = (y[i] - sum) / U[i][i];
  }

  for (int i = 0; i < n; i++) {
    delete[] L[i];
    delete[] U[i];
  }
  delete[] L;
  delete[] U;
  delete[] y;

  SolverResult result;
  result.x = x;
  result.n = n;
  result.iterations = 0;
  result.converged = true;
  result.dominanceAchieved = true;
  result.error = maxError;
  return result;
}

File: Extended_in_python/pynumerics/solvers/doolittle.py
"""
Doolittle LU Decomposition.

Ported from C++ Doolittle class. Decomposes A into L (unit diagonal) and U.
"""

from pynumerics.matrix import Matrix
from pynumerics.solvers.result import SolverResult
from pynumerics.solvers.lu_base import LUDecomposition
from pynumerics.exceptions import MatrixException


class Doolittle(LUDecomposition):
    """LU Decomposition using the Doolittle algorithm (L has unit diagonal)."""

    def solve(self, b: list[float], max_iter: int = 10000, tol: float = 1e-10) -> SolverResult:
        """Solve Ax = b by computing L and U, then substituting.
        
        Args:
            b: Right-hand side vector.
            max_iter: Ignored for direct methods.
            tol: Ignored for direct methods.
            
        Returns:
            SolverResult containing x, L, U, and maximum reconstruction error.
        """
        n = len(b)
        if self.rows != self.cols or self.rows != n:
            raise MatrixException("matrix dimensions don't match for Doolittle LU")

        # Initialize L and U matrices
        L = Matrix(n, n)
        U = Matrix(n, n)
        
        for i in range(n):
            L[i, i] = 1.0

        # Decomposition
        for k in range(n):
            # Upper Triangular
            for j in range(k, n):
                s = sum(L[k, idx] * U[idx, j] for idx in range(k))
                U[k, j] = self._data[k][j] - s
                
            # Lower Triangular
            for i in range(k + 1, n):
                s = sum(L[i, idx] * U[idx, k] for idx in range(k))
                if abs(U[k, k]) < 1e-12:
                    raise MatrixException("zero pivot in Doolittle... matrix might be singular bro")
                L[i, k] = (self._data[i][k] - s) / U[k, k]

        # Calculate max |L*U - A| error
        max_error = 0.0
        for i in range(n):
            for j in range(n):
                s = sum(L[i, k] * U[k, j] for k in range(n))
                err = abs(s - self._data[i][j])
                if err > max_error:
                    max_error = err

        # Forward substitution: Ly = b
        y = [0.0] * n
        for i in range(n):
            s = sum(L[i, j] * y[j] for j in range(i))
            y[i] = b[i] - s

        # Back substitution: Ux = y
        x = [0.0] * n
        for i in range(n - 1, -1, -1):
            s = sum(U[i, j] * x[j] for j in range(i + 1, n))
            if abs(U[i, i]) < 1e-12:
                raise MatrixException("zero on U diagonal in back sub... rip")
            x[i] = (y[i] - s) / U[i, i]

        return SolverResult(
            x=x,
            n=n,
            iterations=0,
            converged=True,
            dominance_achieved=True,
            error=max_error,
            L=L,
            U=U
        )
